### ISSUE TITLE
Eventual Consistency - DataSource `okta_groups` contains destroyed group

### ISSUE THREAD

**Issue by exitcode0** [okta/terraform-provider-okta#1686]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v1.1.6
on darwin_arm64
+ provider registry.terraform.io/hashicorp/aws v3.68.0
+ provider registry.terraform.io/okta/okta v3.41.0

Your version of Terraform is out of date! The latest version
is 1.5.6. You can update by downloading from https://www.terraform.io/downloads.html
```

### Affected Resource(s)

- datasource_okta_groups
- okta_group

### Terraform Configuration Files
```hcl
data "okta_groups" "all_groups" {
  type = "OKTA_GROUP"
}

# resource "okta_group" "aws_123456789012_readOnly {
#   name = "aws_123456789012_readOnly"
# )

resource "okta_app_group_assignments" "aws" {
  app_id = okta_app_saml.aws.id

  dynamic "group" {
    for_each = [
      for group in data.okta_groups.all_groups.groups : group.id if length(regexall("(?i:aws_[0-9]{12})", group.name)) > 0
    ]
    content {
      id      = group.value
      profile = jsonencode({})
    }
  }
}
```

### Expected Behavior

when deleting a `okta_group` resource the datasource `okta_groups` should be tainted or re-evaluated, particularly if it contained said deleted group

### Actual Behavior

it seems that the datasource is either read during the plan or early on in the apply
meaning that if you reference a group from the datasource `okta_groups` by name elsewhere and do something like try to assign it to a app, you'll get a 404

### Steps to Reproduce

see example code, should be pretty straight forward to reproduce

### Important Factoids

Not sure how this should be worked around, we could add delay read to the datasource `okta_groups`
I could also add a `depends_on` or `triggers` in a `null_resource` or some other workaround

perhaps my approach of querying all the groups at once here is the problem and should be avoided
I think I did this to try to save on API requests as the terraform provider is always blowing out rate limits for me

**Comment by monde** [comment]
@exitcode0 I've run into eventual consistency errors in the Okta service in the past. Does your state eventually get consistent with refreshes? If not, we'll use your intuition on the provider side looking for this as a bug.

**Comment by exitcode0** [comment]
The state does become eventually consistent here, in our case a subsequent terraform run went through without issue

Just not sure if I should've done something different here to avoid this

**Comment by monde** [comment]
I'll file an internal bug when we come back to this ticket.

**Comment by duytiennguyen-okta** [comment]
@exitcode0 do you know what is the state of the deleted group is?

**Comment by duytiennguyen-okta** [comment]
OKTA internal reference https://oktainc.atlassian.net/browse/OKTA-644354

**Comment by exitcode0** [comment]
> @exitcode0 do you know what is the state of the deleted group is?

I'm not sure I understand the question
I don't have the state file from this particular instance of hitting this issue,
but i'd imagine it should be fairly easy to repro if its just a terraform graph issue
if its a eventual consistency issue with Okta's API then it might be harder to reproduce

**Comment by duytiennguyen-okta** [comment]
what I meant is when the group is deleted and it still being shown in terraform, what is it status? DELETED? ACTIVE? INACTIVE? If you have the log that would be great

**Comment by monde** [comment]
Right, @exitcode0 if the status is `DELETED` would you expect the provider to filter that group out? If the status is anything other than `DELETED` then I would consider that an eventual consistency issue in the Okta service and we'll need to file a bug on internally at Okta.

**Comment by exitcode0** [comment]
I've looked into this some more
I think there are a few ways the that the behavior I've been seeing could be occurring

An eventual consistency bug in the Okta provider, likely in situations where a resource is read almost immediately after its been deleted (i've not investigated in this space yet)

When a third party system is managing groups outside of Okta (we currently do this)
you can have a DataSource reading groups by name and assigning to a TF managed app
in this scenario its possible for a group to be deleted after the DataSource runs its `read()`  but before the `Apply` (especially if using `Terraform plan -out`
then the `apply` fails when it tries to assign a non existent group from the DataSource to a Okta app


There is also the Scenario outlined below

```hcl
resource "okta_app_bookmark" "google" {
  label = "google"
  url = "https://www.google.com"
  skip_groups = true
  skip_users = true
  lifecycle {
    ignore_changes = [
      users,
      groups
    ]
  }
}
data "okta_groups" "example" { q = "tf_" }
resource "okta_app_group_assignments" "google" {
  app_id = okta_app_bookmark.google.id
  dynamic "group" {
    for_each = [for group in data.okta_groups.example.groups : group.id]
    content {
      id = group.value
    }
  }
}
output "len_groups" {
  value = length(data.okta_groups.example.groups)
}
resource "okta_group" "group_A" {
  name = "tf_group_A"
}
resource "okta_group" "group_B" {
  name = "tf_group_B"
}
# resource "okta_group" "group_C" {
#   name = "tf_group_C"
# }
```
in the above example
1. run `apply -auto-approve`
2. comment `okta_group.group_B`
3. uncomment `okta_group.group_C`
4. run `apply -auto-approve`
 

Perhaps one way to address this is an opt-in option to suppress HTTP 404 `okta_app_group_assignment(s)` resources?
this would stop the plan failure and thing would sort themselves out after the next terraform run

**Comment by exitcode0** [comment]
> Right, @exitcode0 if the status is `DELETED` would you expect the provider to filter that group out? If the status is anything other than `DELETED` then I would consider that an eventual consistency issue in the Okta service and we'll need to file a bug on internally at Okta.

If we're talking about API eventual consistency then I don't think there is an elegant way to solve for this in the provider  
It might be possible to do two `read()` a second or so apart, then compare the pair  
the groups API has a `limit` of like 10k this **_might_** be fine at scale? ¯\\_(ツ)_/¯

As for the other scenarios that I mentioned in the above comment (that have likely muddied the waters on this issue for everyone involved)  
I think the only place to filter these out would be on the consuming resource side, e.g `okta_app_group_assignments`
A resource could read each group before trying to assign 👎  
or we could just suppress 404 as an opt-in flag 👍  
Because Terraform will always do DataSource reads before anything else, I don't think its possible to solve for this on the DataSource itself
AFAIK the Okta Groups API also doesn't return the `Status` of a group, it just returns the Group Object or a 404 response, the Status of the group is implied

TL;DR - I think that adding support for an opt-in flag to suppress `404` responses on a given resource might be the best path forward here

**Comment by exitcode0** [comment]
bumping to remove `waiting-response` label now that https://github.com/okta/terraform-provider-okta/pull/1703 and https://github.com/okta/terraform-provider-okta/pull/1719 have been merged

**Comment by aditya-okta** [comment]
Closing this due to no activity, feel free to reopen this issue if you are still facing it.
